### ISSUE TITLE
dotnet run error Assert failure(PID 28188 [0x00006e1c], Thread: 26664 [0x6828]) in .net 9

### ISSUE THREAD

**Issue by mammadkoma** [dotnet/runtime#110000]
### Description

I create a web api .net 9 project, it builds but after running dotnet run command it shows error. I don't have any problem in .net 8 on Windows 10. 
The commands in .net 9 :
PS D:\TestProjects\WebApi> dotnet build
Restore complete (0.2s)
  WebApi succeeded (2.4s) → bin\Debug\net9.0\WebApi.dll

Build succeeded in 3.0s
PS D:\TestProjects\WebApi> dotnet run
Using launch settings from D:\TestProjects\WebApi\Properties\launchSettings.json...
Building...
CLR: Assert failure(PID 28188 [0x00006e1c], Thread: 26664 [0x6828]): !AreShadowStacksEnabled() || UseSpecialUserModeApc()
    File: D:\a\_work\1\s\src\coreclr\vm\threads.cpp:7938 Image:
D:\TestProjects\WebApi\bin\Debug\net9.0\WebApi.exe
PS D:\TestProjects\WebApi>

### Reproduction Steps

create a web api in .net 9

### Expected behavior

running the project

### Actual behavior

getting error

**Comment by mangod9** [comment]
@mammadkoma this looks related to CET. Are you able to reproduce the failure on different machines? Can you try to disable CET and check if the problem persists: https://learn.microsoft.com/en-us/dotnet/core/compatibility/interop/9.0/cet-support

**Comment by mammadkoma** [comment]
It runs after adding this line to the .csproj file :
<CETCompat>false</CETCompat>

**Comment by mangod9** [comment]
are you able to share a repro? Is this a simple webapi project created with `dotnet new webapi` ?

**Comment by mammadkoma** [comment]
Yes I created by **dotnet new webapi** command.

**Comment by JustArchi** [comment]
Also got user of my OSS app running into the same issue: https://github.com/JustArchiNET/ArchiSteamFarm/issues/3348

Can't provide much more details on their environment, but the app definitely doesn't start now after .NET 9 upgrade.

> CLR: Assert failure(PID 10336 [0x00002860], Thread: 9268 [0x2434]): !AreShadowStacksEnabled() || UseSpecialUserModeApc()
>     File: D:\a\_work\1\s\src\coreclr\vm\threads.cpp:7938 Image:
> D:\program\ASF-win-x64\ArchiSteamFarm.exe

**Comment by mangod9** [comment]
@janvorli FYI. Can they provide details on the hardware they are running on. Also weird that the assert is showing up here. Any repro and/or dumps if available would be helpful to diagnose. 

> So why should I add this line for .net 9 ?

in .NET 9 CET compat-mode is enabled by default, so if the hardware is capable CET would be enabled. It appears there are specific cases where it isnt quite working right.

**Comment by janvorli** [comment]
This means that the Windows are out of date quite a lot and so they don't contain the fix for a bug in the kernel that prevents .NET to work with CET enabled. I believe the fix was released about a year ago.

**Comment by mangod9** [comment]
Can we possibly report a better error here? Not sure how the assert shows up on release builds?

**Comment by MichalPetryka** [comment]
> Can we possibly report a better error here? Not sure how the assert shows up on release builds?

https://github.com/dotnet/runtime/blob/342936c5a88653f0f622e9d6cb727a0e59279b31/src/coreclr/vm/threads.cpp#L7833
It's a release assert.

**Comment by janvorli** [comment]
> Can we possibly report a better error here?

We can replace that assert by a check and failfast with a message that will make more sense.

**Comment by mangod9** [comment]
yeah a better error with some details on required windows versions would be good for better clarity.

**Comment by wwh1004** [comment]
CPU: i7 11800H 
OS: Windows 10 LTSC 21H2 19044.1288 in Hyper-V VM
Repo: https://github.com/dotnet/samples/tree/main/core/extensions/OutOfProcCOM/ExeServer
Publish config: net9.0-windows+win-x64+self-contained

```
C:\Users\admin\Desktop\net9.0-windows>ExeServer.exe /regserver
CLR: Assert failure(PID 4316 [0x000010dc], Thread: 4788 [0x12b4]): !AreShadowStacksEnabled() || UseSpecialUserModeApc()
    File: D:\a\_work\1\s\src\coreclr\vm\threads.cpp:7938 Image:
```

**Comment by mangod9** [comment]
@wwh1004 so looks like you are running an older version of windows which doesnt support the required CET apis.

**Comment by janvorli** [comment]
@wwh1004 I've asked Windows kernel developers and they confirmed that the version of Windows you have doesn't have support for the feature .NET CET support depends on. It was added in an update in 2023. So you need to either disable CET for your app or update your Windows.

**Comment by PJB3005** [comment]
This should probably be documented on the [breaking change page](https://learn.microsoft.com/en-us/dotnet/core/compatibility/interop/9.0/cet-support), as .NET is [still listed](https://github.com/dotnet/core/blob/main/release-notes/9.0/supported-os.md) as being supported on these old Windows versions, and there's no indication that CET needs to be disabled in that scenario.

**Comment by shmuelie** [comment]
I seem to have run into the same issue, but disabling CET didn't fix it. Same error still

**Comment by janvorli** [comment]
@shmuelie when you say it is the same issue, do you mean that you are getting the !AreShadowStacksEnabled() || UseSpecialUserModeApc() assertion failure?

**Comment by shmuelie** [comment]
@janvorli yes, same thing both before and after I set the given property to false

**Comment by janvorli** [comment]
@shmuelie then the CET was not disabled correctly. That assert cannot occur when CET is disabled, as the process doesn't have the shadow stack. How did you try to disable it?

**Comment by janvorli** [comment]
That only works if you publish your app and execute it via the generated .exe file. It doesn't have any effect on the dotnet.exe, so I wonder if you could have used that one to launch your app.

**Comment by shmuelie** [comment]
Ah, that could be it as the tool is being packaged as a .NET Global Tool. Is there any way to work around this for .NET Global tools?

**Comment by mangod9** [comment]
dotnet.exe shouldnt be enabled for CET by default, is the machine enforcing CET for all processes in your case?

**Comment by shmuelie** [comment]
@janvorli @mangod9 Running `Set-ProcessMitigation -Name NameOfTheTool.exe -Disable UserShadowStack` in powershell worked!

**Comment by janvorli** [comment]
> dotnet.exe shouldnt be enabled for CET by default

Right, that was my mistake, I've realized it a bit later that we don't set it so that the dotnet.exe can still run older runtimes that don't have CET support.

**Comment by SamGitCode** [comment]
> [@janvorli](https://github.com/janvorli) [@mangod9](https://github.com/mangod9) Running `Set-ProcessMitigation -Name NameOfTheTool.exe -Disable UserShadowStack` in powershell worked!

Nice trick, this worked for dotnet global tools (which were nuked by this error after their net9 update).

**Comment by mammadkoma** [comment]
After updating to the latest version SDK 9.0.203 it still get the error and I have to add CetCompat false

**Comment by jkotas** [comment]
Updating .NET SDK won't fix this error. This error means that your Windows copy is several years behind on applying updates. The best way to fix this error is by installing all available Windows updates. #114839 updated the error message to say the same.

**Comment by mammadkoma** [comment]
I'm not using the latest version of Windows 10 right now?

![Image](https://github.com/user-attachments/assets/075ca835-9ffa-4854-9b53-40ad80058650)

**Comment by janvorli** [comment]
@mammadkoma the version you have was released in October 11, 2022 (see https://answers.microsoft.com/en-us/insider/forum/all/build-190452130-out-to-release-preview-channel/da795dae-bf69-44e1-8e32-f7c1c6262885). 
The fix was added in 2023.

**Comment by jkotas** [comment]
What's the latest update you see in Windows Update / Update history?

You should see updates getting installed every month: https://support.microsoft.com/en-us/topic/windows-10-update-history-8127c2c6-6edf-4fdf-8b9f-0f7be1ef3562

**Comment by mammadkoma** [comment]
Ok I think I have to update to upper build number.
